pkg/server/biz/scm/bitbucket/bitbucket-server/bitbucket_server.go

```diff
@@ -0,0 +1,422 @@
+	for {
+		files, resp, err := b.v1Client.Repositories.ListFiles(context.Background(), projectKey, repo, &opt)
+		if err != nil {
+			log.Errorf("Fail to list tags for %s", repo)
+			if resp != nil && resp.StatusCode == 500 {
+				return nil, cerr.ErrorSCMServerInternalError.Error(b.scmCfg.Server, err)
+			}
+			return nil, err
+		}
+
+		allFiles = append(allFiles, files.Values...)
+		if files.NextPage == nil {
+			break
+		}
+		opt.Start = files.NextPage
+	}
+
+	var dockerfiles []string
+	for _, file := range allFiles {
+		if strings.HasSuffix(file, "/Dockerfile") {
+			dockerfiles = append(dockerfiles, file)
+		}
+	}
+
+	return dockerfiles, nil
+}
+
+// CreateStatus generate a new status for repository.
+func (b *BitbucketServer) CreateStatus(status c_v1alpha1.StatusPhase, targetURL, repoURL, commitSha string) error {
+	// BitBucket Server:  SUCCESSFUL, FAILED and INPROGRESS.
+	state := "INPROGRESS"
+	description := ""
+
+	switch status {
+	case c_v1alpha1.StatusRunning:
+		state = "INPROGRESS"
+		description = "Cyclone CI is in progress."
+	case c_v1alpha1.StatusSucceeded:
+		state = "SUCCESSFUL"
+		description = "Cyclone CI passed."
+	case c_v1alpha1.StatusFailed:
+		state = "FAILED"
+		description = "Cyclone CI failed."
+	case c_v1alpha1.StatusCancelled:
+		state = "FAILED"
+		description = "Cyclone CI failed."
+	default:
+		err := fmt.Errorf("not supported state:%s", status)
+		log.Error(err)
+		return err
+	}
+
+	label := "continuous-integration/cyclone"
+	opt := &StatusReq{
+		State:       state,
+		Key:         label,
+		Name:        label,
+		URL:         targetURL,
+		Description: description,
+	}
+	_, err := b.v1Client.Repositories.CreateStatus(context.Background(), commitSha, opt)
+	return err
+}
+
+// GetPullRequestSHA gets latest commit SHA of pull request.
+func (b *BitbucketServer) GetPullRequestSHA(repoURL string, number int) (string, error) {
+	projectKey, name := scm.ParseRepo(repoURL)
+	pr, _, err := b.v1Client.PullRequests.GetPullRequest(context.Background(), projectKey, name, number)
+	if err != nil {
+		log.Error(err)
+		return "", err
+	}
+
+	return pr.FromRef.LatestCommit, nil
+}
+
+// CreateWebhook creates webhook for specified repo.
+func (b *BitbucketServer) CreateWebhook(repo string, webhook *scm.Webhook) error {
+	if webhook == nil || len(webhook.URL) == 0 || len(webhook.Events) == 0 {
+		return fmt.Errorf("The webhook %v is not correct", webhook)
+	}
+
+	_, err := b.GetWebhook(repo, webhook.URL)
+	if err != nil && !cerr.ErrorContentNotFound.Derived(err) {
+		return err
+	}
+
+	webhookReq := Webhook{
+		URL:    webhook.URL,
+		Events: make([]string, 0),
+		Active: true,
+		Name:   "continuous-integration/cyclone",
+	}
+
+	for _, e := range webhook.Events {
+		switch e {
+		case scm.PullRequestEventType:
+			webhookReq.Events = append(webhookReq.Events, PrOpened, PrModified)
+		case scm.PullRequestCommentEventType:
+			webhookReq.Events = append(webhookReq.Events, PrCommentAdded)
+		case scm.PushEventType:
```

Comment:
Can merge `scm.PushEventType` and `scm.TagReleaseEventType` in one case.
```suggestion
		case scm.PushEventType, scm.TagReleaseEventType:
```